src/sqlancer/duckdb/test/DuckDBCODDTestOracle.java

```diff
@@ -0,0 +1,1065 @@
+                    .collect(Collectors.toList());
+            dbstate.put(cRef, constants);
+        }
+        this.constantResOfFoldedExpr = new DuckDBResultMap(dbstate, constantRes);
+
+        return select;
+    }
+
+    private void validateExpressionTypes(List<DuckDBCompositeDataType> types) {
+        // float value is inexact, as https://duckdb.org/docs/sql/data_types/numeric#floating-point-types
+        for (DuckDBCompositeDataType type : types) {
+            String name = type.toString();
+            if (name.startsWith("FLOAT") || name.startsWith("DOUBLE") || name.startsWith("REAL")) {
+                throw new IgnoreMeException();
+            }
+        }
+    }
+
+    private List<DuckDBExpression> genJoinExpression(List<DuckDBTableReference> tableList,
+            DuckDBExpression specificCondition) {
+        List<DuckDBExpression> joinExpressions = new ArrayList<>();
+        while (tableList.size() >= 2 && Randomly.getBooleanWithRatherLowProbability()) {
+            DuckDBTableReference leftTable = tableList.remove(0);
+            DuckDBTableReference rightTable = tableList.remove(0);
+            List<DuckDBColumn> columns = new ArrayList<>(leftTable.getTable().getColumns());
+            columns.addAll(rightTable.getTable().getColumns());
+            DuckDBExpressionGenerator joinGen = new DuckDBExpressionGenerator(state).setColumns(columns);
+            DuckDBExpression onPredicate = joinGen.generateExpression();
+            if (specificCondition != null && Randomly.getBooleanWithRatherLowProbability()) {
+                Operator operator = Randomly.fromList(Arrays.asList(DuckDBBinaryComparisonOperator.getRandom(),
+                        DuckDBBinaryArithmeticOperator.getRandom(), DuckDBBinaryLogicalOperator.getRandom()));
+                onPredicate = new DuckDBBinaryOperator(onPredicate, specificCondition, operator);
+            }
+
+            switch (DuckDBJoin.JoinType.getRandom()) {
+            case INNER:
+                joinExpressions.add(DuckDBJoin.createInnerJoin(leftTable, rightTable, joinGen.generateExpression()));
+                break;
+            case NATURAL:
+                joinExpressions.add(DuckDBJoin.createNaturalJoin(leftTable, rightTable, OuterType.getRandom()));
+                break;
+            case LEFT:
+                joinExpressions
+                        .add(DuckDBJoin.createLeftOuterJoin(leftTable, rightTable, joinGen.generateExpression()));
+                break;
+            case RIGHT:
+                joinExpressions
+                        .add(DuckDBJoin.createRightOuterJoin(leftTable, rightTable, joinGen.generateExpression()));
+                break;
+            default:
+                throw new AssertionError();
+            }
+        }
+        return joinExpressions;
+    }
+
+    private List<DuckDBExpression> genOrderBys(DuckDBExpression specificCondition) {
```

Comment:
Do we have similar logic in the expression generator of DuckDB? If so, we can reuse it?